tests/browser/referrences/observe_utils.py

```diff
@@ -0,0 +1,203 @@
+
+    thread = threading.Thread(target=httpd.serve_forever, daemon=True)
+    thread.start()
+
+    try:
+        yield f"http://127.0.0.1:{port}"
+    finally:
+        httpd.shutdown()
+        httpd.server_close()
+        thread.join()
+
+
+def update_json_snapshot() -> list[Path]:  # pragma: no cover
+    """Generate JSON snapshots by automatically starting a temporary server.
+
+    Call this function from a script or CI job whenever you want to refresh all
+    JSON reference files without having to run pytest or start the server
+    manually.
+    """
+
+    # Remove existing JSON and PNG snapshots so the output matches the current
+    # set of HTML files exactly.
+    for json_file in _JSON_DIR.glob("*.json"):
+        try:
+            json_file.unlink()
+        except FileNotFoundError:
+            pass
+
+    with start_local_html_server() as url:
+        return generate_json_snapshot(url)
+
+
+# -----------------------------------------------------------------------------
+# Validation helpers
+# -----------------------------------------------------------------------------
+
+
+def _dump_interaction_nodes(session: "NotteSession") -> list[dict[str, object]]:
+    """Return the serialised interaction nodes for the current session."""
+    nodes_dump: list[dict[str, object]] = []
+    for node in session.snapshot.interaction_nodes():
+        selectors: list[str] = []
+        if node.computed_attributes.selectors is not None:
+            selectors = node.computed_attributes.selectors.selectors()
+
+        nodes_dump.append(
+            {
+                "id": node.id,
+                "role": node.get_role_str(),
+                "text": node.text,
+                "selectors": selectors,
+            }
+        )
+
+    # sort deterministically by id for comparison
+    nodes_dump.sort(key=lambda d: str(d.get("id") or ""))
+    return nodes_dump
+
+
+def check_data_snapshot(base_server_url: str) -> list[str]:
+    """Compare actual nodes extracted from each HTML page against stored JSON.
+
+    Returns a list of names that did *not* match. Empty list means everything
+    is consistent.
+    """
+
+    mismatches: list[str] = []
+
+    for html_file in sorted(_HTML_DIR.glob("*.html")):
+        name = html_file.stem
+        url = f"{base_server_url}/{html_file.name}"
+
+        json_path = _JSON_DIR / f"{name}.json"
+        if not json_path.exists():
+            mismatches.append(name)
+            continue
+
+        with NotteSession(
+            headless=True,
+            enable_perception=False,
+            viewport_width=VIEWPORT_WIDTH,
+            viewport_height=VIEWPORT_HEIGHT,
+        ) as session:  # type: ignore[arg-type]
+            _ = session.observe(url=url)
+
+            actual = _dump_interaction_nodes(session)
+
+        expected = json.loads(json_path.read_text(encoding="utf-8"))
+
+        # sort expected to ensure deterministic order
+        expected.sort(key=lambda d: str(d.get("id") or ""))
```

Comment:
I think it would proably want to check that the selectors are the same. Not the IDs. Because selectors are more stable that our internal notte id systems